File: js-example-app/src/loki.ts
import fetch from "node-fetch";

const loadNs = process.hrtime();
const loadMs = new Date().getTime();

const nanoseconds = () => {
  let diffNs = process.hrtime(loadNs);
  return (loadMs * 1e6 + diffNs[0] * 1e9 + diffNs[1]).toString();
};

export class LokiClient {
  constructor(
    private readonly url: string,
    private readonly labels: Record<string, string>
  ) {}

  private createLokiInput(labels: Record<string, string>, message: string) {
    return JSON.stringify({
      streams: [
        {
          stream: labels,
          values: [[nanoseconds(), message]],
        },
      ],
    });
  }

  async sendLog(message: string) {
    const body = this.createLokiInput(this.labels, message);
    return fetch(this.url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body,
    });
  }
}

File: python-example-app/src/loki.py
import json
from time import time_ns

import aiohttp


class LokiClient:
    def __init__(self, url: str, labels: dict[str, str]):
        self._url = url
        self._labels = labels

    def create_loki_input(self, labels: dict[str, str], message: str):
        return dict(streams=[dict(stream=labels, values=[[str(time_ns()), message]])])

    async def send_log(self, message: str) -> tuple[str, int]:
        body = self.create_loki_input(self._labels, message)

        async with aiohttp.ClientSession() as session:
            async with session.post(
                self._url, data=json.dumps(body), headers={"Content-Type": "application/json"}
            ) as response:
                return await response.text(), response.status
